edb/graphql/translator.py

```diff
@@ -1612,15 +1612,42 @@ def _visit_list_of_inputs(self, inputlist, op):
             raise g_errors.GraphQLTranslationError(
                 f'a list was expected')
 
-        result = [self.visit(node) for node in inputlist.values]
-        return self._join_expressions(result, op)
-
-    def _join_expressions(self, exprs, op='AND'):
-        if not exprs:
+        return self._visit_list_generalized_bool_op(
+            [node for node in inputlist.values], op)
+
+    def _visit_list_generalized_bool_op(self, nodes, op):
+        # Generalization of a boolean operation AND or OR as it is
+        # applied to a list of expressions. This comes up in filters
+        # either explicitly by using 'and' or 'or' or by supplying a
+        # list of expressions where 'and' is implied.
+        #
+        # In this limited context it is appropriate to use Postres'
+        # truth table for AND and OR, short-circuiting "False AND
+        # anything" or "True OR anything" respectively to "False" and
+        # "True" instead of the stricter EdgeQL rules that would
+        # produce empty sets if any of the inputs are empty.
+
+        if not nodes:
             return None
-        elif len(exprs) == 1:
-            return exprs[0]
+        elif len(nodes) == 1:
+            return self.visit(nodes[0])
+
+        # The short-circuiting value is True for OR and False for AND.
+        short = op == 'OR'
+
+        return qlast.IfElse(
```

Comment:
I don't think using `IfElse` would be friendly to the query plan, it's very likely Postgres wouldn't be able to use bitmap index scans if a `CASE WHEN` is involved.  I'm afraid we don't have a choice but to introduce the short-circuiting versions of `AND` and `OR`, possibly as functions.

Reply:
I've added `sys::__pg_and` and `sys::__pg_or` functions marked as "internal".